File: leetcode_75/Binary_Tree_DFS/cpp/872.cpp
#include <vector>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int get_leaves(TreeNode *node, vector<int> &leaves)
    {
        if (node == nullptr)
        {
            return 0;
        }
        if (node->left == nullptr && node->right == nullptr)
        {
            leaves.push_back(node->val);
            return 0;
        }
        return this->get_leaves(node->left, leaves) + this->get_leaves(node->right, leaves);
    }

    bool leafSimilar(TreeNode *root1, TreeNode *root2)
    {
        vector<int> t1;
        vector<int> t2;
        this->get_leaves(root1, t1);
        this->get_leaves(root2, t2);

        return t1 == t2;
    }
};

File: leetcode_75/Binary_Tree_DFS/python/872.py
from typing import Optional
from functools import *


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    @lru_cache(None)
    def get_leaves(self, node: TreeNode):
        if node is None:
            return []
        if node.left is None and node.right is None:
            return [node.val]
        return self.get_leaves(node.left) + self.get_leaves(node.right)

    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        return self.get_leaves(root1) == self.get_leaves(root2)
